doc/neps/nep-0018-array-function-protocol.rst

```diff
@@ -199,6 +200,68 @@ include *all* of the corresponding NumPy function's optional arguments
 were explicitly used in the NumPy function call.
 
+.. note::
+
+    Just like the case for builtin special methods like ``__add__``, properly
+    written ``__array_function__`` methods should always return
+    ``NotImplemented`` when an unknown type is encountered. Otherwise, it will
+    be impossible to correctly overrides NumPy functions from another object
+    if the operation also includes one of your objects.
+
+Defaulting to NumPy's implementation
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Some projects may prefer to default to NumPy's implementation, rather than
+explicitly defining implementing a supported API. This allows for incrementally
+overriding NumPy's API in projects that already support it implicitly by
+allowing their objects to be converted into NumPy arrays (e.g., because they
+implemented special methods such as ``__array__``). We don't recommend this
+for most new projects ("Explicit is better than implicit"), but in some cases
+it is the most expedient option.
+
+This is possible by making use of the special ``__numpy_implementation__``
+attribute found on NumPy functions that that support overrides with
+``__array_function__``. ``__numpy_implementation__`` always points back to
+the original NumPy-array specific implementation of a function. These functions
+do not check for a overrides, and instead usually coerce all of their
+array-like arguments to NumPy arrays.
+
+Adapting the previous example:
+
+.. code:: python
+
+    class MyArray:
+        def __array_function__(self, func, types, args, kwargs):
+            # It is still best practice to defer to unrecognized types
+            if not all(issubclass(t, (MyArray, np.ndarray)) for t in types):
+                return NotImplemented
+
+            my_func = HANDLED_FUNCTIONS.get(func)
+            if my_func is None:
+                return func.__numpy_implementation__(*args, **kwargs)
+            return my_func(*args, **kwargs)
+
+        def __array__(self, dtype):
+            # convert this object into a NumPy array
+
+Now, if a NumPy function that isn't explicitly handled is called on
+``MyArray`` object, the operation will act (almost) as if MyArray's
+``__array_function__`` method never existed.
+
+The one difference in behavior is in cases where multiple objects implement
+``__array_function__`` but don't know about each other: NumPy will raise
+``TypeError`` if all methods return ``NotImplemented``, whereas if no arguments
+defined ``__array_function__`` methods it would attempt to coerce all of them
+to NumPy arrays. Of course, you could "fix" this for your class by never
+returning ``NotImplemented`` from your ``__array_function__`` method, but in
+general we think this is a good change: fallback implementations often result
+in unpredictable and undesired behavior.
+
+.. note::
+
+    ``__numpy_implementation__`` was not included as part of the initial opt-in
+    only preview of ``__array_function__`` in NumPy 1.16.
```

Comment:
nit: I think this needs to be hyphenated as `opt-in-only`

Reply:
Actually, I once got a manuscript back (in the time the journals still did it), where I was told that that is one of the purposes of a double hyphen: opt--in only. I must admit it still reads strange to me, but for really long ones it does make sense.

Reply:
https://www.dailywritingtips.com/5-examples-of-the-need-for-multiple-hyphenation/

Reply:
Astrophysical Journal does not follow that - but I do find the examples convincing.